### PR TITLE
WIP: Feat folders

### PR DESCRIPTION
**Pull request by seavan**

### CODE REVIEW

File: src/stores/ui-store.js [hunk 1]
```diff
@@ -7,6 +7,10 @@ const appControl = require('~/helpers/app-control');
 class UIStore {
     @observable contactDialogUsername;
 
+    // current folder selected in Files.jsx
+    // used by drop
+    @observable currentFolder;
```
Comment (anri-asaturov) [inline src/stores/ui-store.js]: Why not in SDK, @seavan ?

Comment (seavan) [reply src/stores/ui-store.js]: are you sure we need it in SDK? currentFolder is used by Files.jsx component, but not used by "move to folder" dialog, for example.

Comment (anri-asaturov) [reply src/stores/ui-store.js]: Well that's what i was thinking, maybe fileStore.curentFolder needs to be a thing and everyone (sdk, desktop, mobile) uses it, but i don't insist, might not be a good idea

File: src/ui/chat/components/sidebar/FilesSection.jsx [hunk 2]
```diff
@@ -62,15 +62,44 @@ class FilesSection extends React.Component {
...
+                                >
+                                    <span className="rt-list-item">
+                                        <span className="rt-list-itemContentRoot">
+                                            <span className="rt-list-itemText rt-list-primary">
+                                                <span className="sidebar-file-label">
+                                                    <FileSpriteIcon type={file.iconType} size="medium" />
+                                                    <div className="meta">
+                                                        <div className="file-name-container">
+                                                            <span className="file-name">
+                                                                {file.nameWithoutExtension}
```
Comment (anri-asaturov) [inline src/ui/chat/components/sidebar/FilesSection.jsx]: Let's extract at least internal render function from something here to reduce nesting.

File: src/ui/chat/components/sidebar/FilesSection.jsx [hunk 3]
```diff
@@ -62,15 +62,44 @@ class FilesSection extends React.Component {
...
+                                                            </span>
+                                                            <span className="file-ext">.{file.ext}</span>
+                                                        </div>
+                                                        <span className="file-shared-by">
+                                                            {file.fileOwner === User.current.username
+                                                                ? `${t('title_fileFilterShared')}`
+                                                                : `${t('title_fileSharedByUser',
+                                                                    { user: file.fileOwner })}`
```
Comment (anri-asaturov) [inline src/ui/chat/components/sidebar/FilesSection.jsx]: This needs to be T component, parametrized locales create react key warnings

File: src/ui/files/Files.jsx [hunk 4]
```diff
@@ -18,10 +23,13 @@ class Files extends React.Component {
         this.handleUpload = this.handleUpload.bind(this);
     }
 
-    @observable renderedItemsCount = 15;
-    pageSize = 15;
+    @observable renderedItemsCount = DEFAULT_RENDERED_ITEMS_COUNT;
+    pageSize = DEFAULT_RENDERED_ITEMS_COUNT;
 
     componentWillMount() {
+        if (!uiStore.currentFolder) {
```
Comment (anri-asaturov) [inline src/ui/files/Files.jsx]: what if currentFolder gets deleted (on other device and syncs with this one) when we render this, btw, does something account for that?

File: src/ui/files/Files.jsx [hunk 5]
```diff
@@ -32,16 +40,161 @@ class Files extends React.Component {
...
+        if (val === '') {
+            fileStore.clearFilter();
+            return;
+        }
+        fileStore.filterByName(val);
+    };
+
+    @observable moveFolderVisible = false;
+
+    @action moveFolder = folder => {
```
Comment (anri-asaturov) [inline src/ui/files/Files.jsx]: New rules! This should be `@action.bound` and every other action that gets assigned to arrow function in this PR

File: src/ui/files/Files.jsx [hunk 6]
```diff
@@ -32,16 +40,161 @@ class Files extends React.Component {
...
+        this.folderToMove = folder;
+        this.moveFolderVisible = true;
+    }
+
+    hideMoveFolder = () => {
```
Comment (anri-asaturov) [inline src/ui/files/Files.jsx]: this should be a bound action

File: src/ui/files/Files.jsx [hunk 7]
```diff
@@ -32,16 +40,161 @@ class Files extends React.Component {
...
+        await Promise.all(list.success.map(path => {
+            const file = fileStore.upload(
+                path, null, uiStore.currentFolder.isRoot ? null : uiStore.currentFolder.folderId
+            );
+            return new Promise(resolve =>
+                when(() => file.fileId, () => {
+                    uiStore.currentFolder.moveInto(file);
+                    resolve();
+                }));
+        }));
```
Comment (anri-asaturov) [inline src/ui/files/Files.jsx]: This kinda stuff needs to be in sdk tbh

File: src/ui/files/Files.jsx [hunk 8]
```diff
@@ -101,44 +255,110 @@ class Files extends React.Component {
...
+    }
+
+    get breadCrumbsHeader() {
+        return (
+            <div className="files-header">
+                <Breadcrumb currentFolder={uiStore.currentFolder}
+                    onSelectFolder={this.changeFolder}
+                    onMove={() => this.moveFolder(uiStore.currentFolder)}
+                    onDelete={() => this.deleteFolder(uiStore.currentFolder)}
+                    onRename={() => this.showRenameFolderPopup(uiStore.currentFolder)}
```
Comment (anri-asaturov) [inline src/ui/files/Files.jsx]: Refactor arrow handlers.

File: src/ui/files/components/Breadcrumb.jsx [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        let iterator = this.props.currentFolder;
+        do {
+            folderPath.unshift(iterator);
+            iterator = iterator.parent;
+        } while (iterator);
+        return folderPath;
+    }
+
+    componentDidMount() {
+        reaction(() => this.props.currentFolder, this.ellipsizeFolderNames, true);
```
Comment (anri-asaturov) [inline src/ui/files/components/Breadcrumb.jsx]: destroy reaction on component unmount

File: src/ui/files/components/Breadcrumb.jsx [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+    render() {
+        const { folderPath } = this;
+        return (
+            <div className="breadcrumb-container">
+                <div className="breadcrumb">
+                    {this.folderPath.map((folder, i) => (
+                        <div key={`${i}-${folder.name}` || 'root'} className="breadcrumb-entry">
+                            <a className="folder-link clickable"
+                                onClick={() => this.handleClick(folder)}>
```
Comment (anri-asaturov) [inline src/ui/files/components/Breadcrumb.jsx]: arrow handler

File: src/ui/files/components/FilePicker.jsx [hunk 11]
```diff
@@ -44,40 +102,67 @@ class FilePicker extends React.Component {
...
+                    onChangeFolder={() => this.changeFolder(f)}
+                    checkboxPlaceholder
+                /> :
+                <FileLine
+                    key={f.fileId}
+                    file={f}
+                    currentFolder={currentFolder}
+                    checkbox
+                    selected={f.selected}
+                    onToggleSelect={() => this.toggleSelect(f)}
```
Comment (anri-asaturov) [inline src/ui/files/components/FilePicker.jsx]: arrow handlers, here and line 114

File: src/ui/files/components/FolderLine.jsx [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+                {this.props.checkboxPlaceholder &&
+                    <div className="checkbox" />
+                }
+
+                <div className="file-icon">
+                    <FontIcon value="folder" />
+                </div>
+
+                <div className="file-name clickable selectable"
+                    onClick={() => this.props.onChangeFolder(this.props.folder)} >
```
Comment (anri-asaturov) [inline src/ui/files/components/FolderLine.jsx]: arrow handler

File: src/ui/files/components/FolderLine.jsx [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+                {this.props.folderDetails && <div className="file-uploaded text-right" /> }
+
+                {this.props.folderDetails && <div className="file-size text-right" /> }
+
+                {this.props.folderActions &&
+                    <div className="file-actions text-right">
+                        <FolderActions
+                            onRename={() => this.props.onRenameFolder(this.props.folder)}
+                            onMove={() => this.props.onMoveFolder(this.props.folder)}
+                            onDelete={() => this.props.onDeleteFolder(this.props.folder)}
```
Comment (anri-asaturov) [inline src/ui/files/components/FolderLine.jsx]: arrow handlers

File: src/ui/files/components/MoveFileDialog.jsx [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+            { label: t('button_cancel'), onClick: this.props.onHide },
+            { label: t('button_move'), onClick: this.handleMove }
+        ];
+
+        const folders = currentFolder.folders.filter(f => f !== this.props.folder).map(folder => (
+            <div key={`folder-${folder.folderId}`} className="move-file-row">
+                <Button
+                    icon={this.selectedFolder === folder ?
+                        'radio_button_checked' : 'radio_button_unchecked'}
+                    onClick={() => this.selectionChange(folder)}
```
Comment (anri-asaturov) [inline src/ui/files/components/MoveFileDialog.jsx]: arrow handler (here and some others below)

File: src/ui/chat/components/sidebar/FilesSection.jsx (hunk 3 above)
Comment (seavan) [reply src/ui/chat/components/sidebar/FilesSection.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/5e98f6ae100fc0922d946211843d7a4caa98ef72

File: src/ui/chat/components/sidebar/FilesSection.jsx (hunk 2 above)
Comment (seavan) [reply src/ui/chat/components/sidebar/FilesSection.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/21a424efe8a85028786201b1b49934fcab1399be

File: src/ui/files/Files.jsx (hunk 5 above)
Comment (seavan) [reply src/ui/files/Files.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/fb15011fc02f61a7311c302c1f349a5c54f395bc

File: src/ui/files/Files.jsx (hunk 6 above)
Comment (seavan) [reply src/ui/files/Files.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/fb15011fc02f61a7311c302c1f349a5c54f395bc

File: src/ui/files/components/Breadcrumb.jsx (hunk 10 above)
Comment (seavan) [reply src/ui/files/components/Breadcrumb.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/4ab63e37d355a6ea783a7e1a149b6ecbbdd51f63

File: src/ui/files/components/FolderLine.jsx (hunk 12 above)
Comment (seavan) [reply src/ui/files/components/FolderLine.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/4ab63e37d355a6ea783a7e1a149b6ecbbdd51f63

File: src/ui/files/components/FolderLine.jsx (hunk 13 above)
Comment (seavan) [reply src/ui/files/components/FolderLine.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/4ab63e37d355a6ea783a7e1a149b6ecbbdd51f63

File: src/ui/files/Files.jsx (hunk 8 above)
Comment (seavan) [reply src/ui/files/Files.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/4ab63e37d355a6ea783a7e1a149b6ecbbdd51f63

File: src/ui/files/components/Breadcrumb.jsx (hunk 9 above)
Comment (seavan) [reply src/ui/files/components/Breadcrumb.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/33841555b0fe00738a98b429c5a9e134e6c02718

File: src/ui/files/components/FilePicker.jsx (hunk 11 above)
Comment (seavan) [reply src/ui/files/components/FilePicker.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/5cf23b7bac5e72594c4788b507bba4fefdeb00ed

File: src/ui/files/components/MoveFileDialog.jsx (hunk 14 above)
Comment (seavan) [reply src/ui/files/components/MoveFileDialog.jsx]: already refactored in merged PR

File: src/stores/ui-store.js (hunk 1 above)
Comment (seavan) [reply src/stores/ui-store.js]: it makes sense to move currentFolder for SDK at least for Files.jsx (maybe not for FilePicker so it doesn't affect the global status). did it here:
https://github.com/PeerioTechnologies/peerio-desktop/commit/35f34c12e9e3c0f4bbc6c92d6f834eb3004cc33b
https://github.com/PeerioTechnologies/peerio-icebear/commit/bb77f9518da9d0459153f1b072ec9fb43937ac7f

File: src/ui/files/Files.jsx (hunk 4 above)
Comment (seavan) [reply src/ui/files/Files.jsx]: fixed in here:

https://github.com/PeerioTechnologies/peerio-desktop/commit/35f34c12e9e3c0f4bbc6c92d6f834eb3004cc33b
https://github.com/PeerioTechnologies/peerio-icebear/commit/bb77f9518da9d0459153f1b072ec9fb43937ac7f

File: src/ui/files/Files.jsx (hunk 7 above)
Comment (seavan) [reply src/ui/files/Files.jsx]: https://github.com/PeerioTechnologies/peerio-desktop/commit/dc7f19d6364a20188f7745174c44b3c042bdcd62
https://github.com/PeerioTechnologies/peerio-icebear/commit/80dbf90188b1bb664aa95334deb69f4ff53e75eb

**Pull request merged**